accel/tcg/log_instr.c

```diff
@@ -0,0 +1,513 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2020 Alfredo Mazzinghi
+ * All rights reserved.
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory under DARPA/AFRL contract FA8750-10-C-0237
+ * ("CTSRD"), as part of the DARPA CRASH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <stdint.h>
+
+#include "qemu/osdep.h"
+#include "qemu/log.h"
+#include "cpu.h"
+#include "exec/log.h"
+#include "exec/helper-proto.h"
+#include "exec/log_instr.h"
+#include "disas/disas.h"
+
+/*
+ * CHERI common instruction logging.
+ *
+ * This is the central implementation of the CPU_LOG_INSTR, CPU_LOG_CVTRACE
+ * and CPU_LOG_USER_ONLY methods.
+ * The same functions can be used by CHERI targets to append to the instruction
+ * log buffer. Once the instruction is fully processed, the target commits the
+ * log buffer and depending on the instruction operations and -dfilter options
+ * we either flush the buffer or drop it.
+ * A central design goal is to deterministically output the main side effect of
+ * the instruction (generally the destination register value), other side
+ * effects, such as multiple register updates, may be implemented by emitting
+ * multiple trace entries, but this is currently unimplemented.
+ * Another design goal is to allow tracing of arbitrary events via nops in
+ * both the text and binary format.
+ */
+
+#ifdef CONFIG_CHERI_LOG_INSTR
+
+#ifdef DEBUG_INSTR_LOGGING
```

Comment:
Should this automatically be on if CONFIG_DEBUG_TCG is on? DEBUG_TCG builds are already extremely slow, adding more assertions probably doesn't matter.

Reply:
Good point.